TiFlash crashed when PageStorage V3 restore

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Run ZTO workload, kill TiFlash process randomly.
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
TiFlash can restart and the ZTO workload can be executed successfully after TiFlash restart.
### 3. What did you see instead (Required)
TiFlash crashed when PageStorage V3 restore.
`[2022/05/09 23:44:15.381 +08:00] [ERROR] [Exception.cpp:85] ["DB::EngineStoreApplyRes DB::HandleAdminRaftCmd(const DB::EngineStoreServerWrap *, DB::BaseBuffView, DB::BaseBuffView, DB::RaftCmdHeader):Code: 0, e.displayText() = DB::Exception: Invalid page id [page_id=90.2018] [resolve_id=90.2010], e.what() = DB::Exception, Stack trace:          0x1d082a3\tStackTrace::StackTrace() [tiflash+30442147]                 \tdbms/src/Common/StackTrace.cpp:23        0x1d058a6\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+30431398]                 \tdbms/src/Common/Exception.h:41        0x7a3db20\tDB::PS::V3::PageDirectory::getNormalPageId(DB::UInt128, std::__1::shared_ptr<DB::PS::V3::PageDirectorySnapshot> const&, bool) const [tiflash+128178976]                 \tdbms/src/Storages/Page/V3/PageDirectory.cpp:830        0x7a16f1a\tDB::PS::V3::PageStorageImpl::getNormalPageIdImpl(unsigned long, unsigned long, std::__1::shared_ptr<DB::PageStorageSnapshot>, bool) [tiflash+128020250]                 \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:76        0x77f2091\tDB::PageReader::getNormalPageId(unsigned long) const [tiflash+125771921]                 \tdbms/src/Storages/Page/PageStorage.h:372        0x785ff61\tDB::DM::ColumnFileBig::deserializeMetadata(DB::DM::DMContext&, DB::DM::RowKeyRange const&, DB::ReadBuffer&) [tiflash+126222177]                 \tdbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileBig.cpp:75        0x7881f53\tDB::DM::deserializeSavedColumnFilesInV3Format(DB::DM::DMContext&, DB::DM::RowKeyRange const&, DB::ReadBuffer&, unsigned long) [tiflash+126361427]                 \tdbms/src/Storages/DeltaMerge/ColumnFile/ColumnFile_V3.cpp:86        0x787901a\tDB::DM::deserializeSavedColumnFiles(DB::DM::DMContext&, DB::DM::RowKeyRange const&, DB::ReadBuffer&) [tiflash+126324762]                 \tdbms/src/Storages/DeltaMerge/ColumnFile/ColumnFilePersisted.cpp:123        0x788add1\tDB::DM::ColumnFilePersistedSet::restore(DB::DM::DMContext&, DB::DM::RowKeyRange const&, unsigned long) [tiflash+126397905]                 \tdbms/src/Storages/DeltaMerge/Delta/ColumnFilePersistedSet.cpp:112        0x78849ea\tDB::DM::DeltaValueSpace::restore(DB::DM::DMContext&, DB::DM::RowKeyRange const&, unsigned long) [tiflash+126372330]                 \tdbms/src/Storages/DeltaMerge/Delta/DeltaValueSpace.cpp:59        0x77a549e\tDB::DM::Segment::restoreSegment(DB::DM::DMContext&, unsigned long) [tiflash+125457566]                 \tdbms/src/Storages/DeltaMerge/Segment.cpp:274        0x776ba0c\tDB::DM::DeltaMergeStore::DeltaMergeStore(DB::Context&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, long, std::__1::vector<DB::DM::ColumnDefine, std::__1::allocator<DB::DM::ColumnDefine> > const&, DB::DM::ColumnDefine const&, bool, unsigned long, DB::DM::DeltaMergeStore::Settings const&) [tiflash+125221388]                 \tdbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:263        0x76f706d\tDB::StorageDeltaMerge::getAndMaybeInitStore() [tiflash+124743789]                 \tdbms/src/Storages/StorageDeltaMerge.cpp:1527        0x76f9902\tDB::StorageDeltaMerge::flushCache(DB::Context const&, DB::DM::RowKeyRange const&) [tiflash+124754178]                 \tdbms/src/Storages/StorageDeltaMerge.cpp:773        0x78dc4e2\tDB::KVStore::tryFlushRegionCacheInStorage(DB::TMTContext&, DB::Region const&, Poco::Logger*) [tiflash+126731490]                 \tdbms/src/Storages/Transaction/KVStore.cpp:153        0x78deb5f\tDB::KVStore::handleUselessAdminRaftCmd(raft_cmdpb::AdminCmdType, unsigned long, unsigned long, unsigned long, DB::TMTContext&) [tiflash+126741343]                 \tdbms/src/Storages/Transaction/KVStore.cpp:384        0x78ded3c\tDB::KVStore::handleAdminRaftCmd(raft_cmdpb::AdminRequest&&, raft_cmdpb::AdminResponse&&, unsigned long, unsigned long, unsigned long, DB::TMTContext&) [tiflash+126741820]                 \tdbms/src/Storages/Transaction/KVStore.cpp:411        0x78f7685\tHandleAdminRaftCmd [tiflash+126842501]                 \tdbms/src/Storages/Transaction/ProxyFFI.cpp:114   0x7f4c3fb801de\traftstore::engine_store_ffi::_$LT$impl$u20$raftstore..engine_store_ffi..interfaces..root..DB..EngineStoreServerHelper$GT$::handle_admin_raft_cmd::hf504663a656e474e [libtiflash_proxy.so+21152222]   0x7f4c4031d669\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::exec_admin_cmd::hcd543b3e2a90c2c5 [libtiflash_proxy.so+29136489]   0x7f4c403109cf\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::process_raft_cmd::hf45229906fb926c9 [libtiflash_proxy.so+29084111]   0x7f4c4030f305\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::handle_raft_committed_entries::ha4dbdc3a3458dcda [libtiflash_proxy.so+29078277]   0x7f4c403303cd\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_apply::h574db8b558a21550 [libtiflash_proxy.so+29213645]   0x7f4c403332a9\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_tasks::h5316fcd72343e9be [libtiflash_proxy.so+29225641]   0x7f4c40335cb0\t_$LT$raftstore..store..fsm..apply..ApplyPoller$LT$EK$GT$$u20$as$u20$batch_system..batch..PollHandler$LT$raftstore..store..fsm..apply..ApplyFsm$LT$EK$GT$$C$raftstore..store..fsm..apply..ControlFsm$GT$$GT$::handle_normal::hec3475cd3a78b2fd [libtiflash_proxy.so+29236400]   0x7f4c3ffa2365\tbatch_system::batch::Poller$LT$N$C$C$C$Handler$GT$::poll::h5af46678afa583e3 [libtiflash_proxy.so+25486181]   0x7f4c406689ab\tstd::sys_common::backtrace::__rust_begin_short_backtrace::h0b8b7288f3f90d60 [libtiflash_proxy.so+32590251]   0x7f4c3fe8e62c\tcore::ops::function::FnOnce::call_once$u7b$$
`
### 4. What is your TiFlash version? (Required)
8781b0fcb3394ebd7817fd1ddab713021076d242 or 8781b0fcb3394ebd7817fd1ddab713021076d242 with PR https://github.com/pingcap/tiflash/pull/4850

--- a/dbms/src/Storages/Page/V3/PageDirectory.cpp
+++ b/dbms/src/Storages/Page/V3/PageDirectory.cpp
@@ -334,6 +334,11 @@ VersionedPageEntries::resolveToPageId(UInt64 seq, bool check_prev, PageEntryV3 *
             return {RESOLVE_TO_REF, ori_page_id, create_ver};
         }
     }
+    else
+    {
+        LOG_FMT_WARNING(&Poco::Logger::get("VersionedPageEntries"), "Can't reslove the EditRecordType {}", type);
+    }
+
     return {RESOLVE_FAIL, buildV3Id(0, 0), PageVersionType(0)};
 }
 

--- a/dbms/src/Storages/Page/V3/PageDirectory.cpp
+++ b/dbms/src/Storages/Page/V3/PageDirectory.cpp
@@ -721,6 +721,11 @@ PageIDAndEntryV3 PageDirectory::get(PageIdV3Internal page_id, const PageDirector
             {
                 if (throw_on_not_exist)
                 {
+                    LOG_FMT_WARNING(log, "Dump state for invalid page id [page_id={}]", page_id);
+                    for (const auto &[dump_id, dump_entry]: mvcc_table_directory)
+                    {
+                        LOG_FMT_WARNING(log, "Dumping state [page_id={}] [entry={}]", dump_id, dump_entry == nullptr ? "<null>" : dump_entry->toDebugString());
+                    }
                     throw Exception(fmt::format("Invalid page id, entry not exist [page_id={}] [resolve_id={}]", page_id, id_to_resolve), ErrorCodes::PS_ENTRY_NOT_EXISTS);
                 }
                 else

--- a/dbms/src/Storages/Page/V3/PageDirectory.cpp
+++ b/dbms/src/Storages/Page/V3/PageDirectory.cpp
@@ -722,7 +722,7 @@ PageIDAndEntryV3 PageDirectory::get(PageIdV3Internal page_id, const PageDirector
                 if (throw_on_not_exist)
                 {
                     LOG_FMT_WARNING(log, "Dump state for invalid page id [page_id={}]", page_id);
-                    for (const auto &[dump_id, dump_entry]: mvcc_table_directory)
+                    for (const auto & [dump_id, dump_entry] : mvcc_table_directory)
                     {
                         LOG_FMT_WARNING(log, "Dumping state [page_id={}] [entry={}]", dump_id, dump_entry == nullptr ? "<null>" : dump_entry->toDebugString());
                     }

--- a/dbms/src/Storages/Page/V3/WAL/serialize.cpp
+++ b/dbms/src/Storages/Page/V3/WAL/serialize.cpp
@@ -100,7 +100,6 @@ void deserializePutFrom([[maybe_unused]] const EditRecordType record_type, ReadB
     UInt32 flags = 0;
     readIntBinary(flags, buf);
 
-    // All consider as put
     PageEntriesEdit::EditRecord rec;
     rec.type = record_type;
     readIntBinary(rec.page_id, buf);

--- a/dbms/src/Storages/Page/V3/WAL/serialize.cpp
+++ b/dbms/src/Storages/Page/V3/WAL/serialize.cpp
@@ -152,7 +152,7 @@ void deserializePutExternalFrom([[maybe_unused]] const EditRecordType record_typ
     assert(record_type == EditRecordType::PUT_EXTERNAL || record_type == EditRecordType::VAR_EXTERNAL);
 
     PageEntriesEdit::EditRecord rec;
-    rec.type = EditRecordType::PUT_EXTERNAL;
+    rec.type = record_type;
     readIntBinary(rec.page_id, buf);
     deserializeVersionFrom(buf, rec.version);
     readIntBinary(rec.being_ref_count, buf);

--- a/dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp
@@ -23,6 +23,7 @@
 #include <Storages/Page/V3/PageDirectoryFactory.h>
 #include <Storages/Page/V3/PageEntriesEdit.h>
 #include <Storages/Page/V3/PageEntry.h>
+#include <Storages/Page/V3/WAL/serialize.h>
 #include <Storages/Page/V3/WALStore.h>
 #include <Storages/Page/V3/tests/entries_helper.h>
 #include <Storages/tests/TiFlashStorageTestBasic.h>
@@ -1760,16 +1761,88 @@ try
 }
 CATCH
 
-TEST_F(PageDirectoryGCTest, DumpAndRestore)
+
+TEST_F(PageDirectoryGCTest, GCOnRefedExternalEntries2)
 try
 {
+    {
+        PageEntriesEdit edit; // ingest
+        edit.putExternal(352);
+        dir->apply(std::move(edit));
+    }
+    {
+        PageEntriesEdit edit;
+        edit.ref(353, 352);
+        dir->apply(std::move(edit));
+    }
+    {
+        PageEntriesEdit edit; // ingest done
+        edit.del(352);
+        dir->apply(std::move(edit));
+    }
+    {
+        PageEntriesEdit edit; // split
+        edit.ref(357, 352);
+        edit.ref(359, 352);
+        dir->apply(std::move(edit));
+    }
+    {
+        PageEntriesEdit edit; // split done
+        edit.del(353);
+        dir->apply(std::move(edit));
+    }
+    {
+        PageEntriesEdit edit; // one of segment delta-merge
+        edit.del(359);
+        dir->apply(std::move(edit));
+    }
+
+    {
+        auto snap = dir->createSnapshot();
+        auto normal_id = dir->getNormalPageId(357, snap);
+        EXPECT_EQ(normal_id.low, 352);
+    }
+    dir->gcInMemEntries();
+    {
+        auto snap = dir->createSnapshot();
+        auto normal_id = dir->getNormalPageId(357, snap);
+        EXPECT_EQ(normal_id.low, 352);
+    }
+
+    auto s0 = dir->createSnapshot();
+    auto edit = dir->dumpSnapshotToEdit(s0);
+    edit.size();
     auto restore_from_edit = [](const PageEntriesEdit & edit) {
+        auto deseri_edit = DB::PS::V3::ser::deserializeFrom(DB::PS::V3::ser::serializeTo(edit));
         auto ctx = DB::tests::TiFlashTestEnv::getContext();
         auto provider = ctx.getFileProvider();
         auto path = getTemporaryPath();
         PSDiskDelegatorPtr delegator = std::make_shared<DB::tests::MockDiskDelegatorSingle>(path);
         PageDirectoryFactory factory;
-        auto d = factory.createFromEdit(getCurrentTestName(), provider, delegator, edit);
+        auto d = factory.createFromEdit(getCurrentTestName(), provider, delegator, deseri_edit);
+        return d;
+    };
+    {
+        auto restored_dir = restore_from_edit(edit);
+        auto snap = restored_dir->createSnapshot();
+        auto normal_id = restored_dir->getNormalPageId(357, snap);
+        EXPECT_EQ(normal_id.low, 352);
+    }
+}
+CATCH
+
+
+TEST_F(PageDirectoryGCTest, DumpAndRestore)
+try
+{
+    auto restore_from_edit = [](const PageEntriesEdit & edit) {
+        auto deseri_edit = DB::PS::V3::ser::deserializeFrom(DB::PS::V3::ser::serializeTo(edit));
+        auto ctx = DB::tests::TiFlashTestEnv::getContext();
+        auto provider = ctx.getFileProvider();
+        auto path = getTemporaryPath();
+        PSDiskDelegatorPtr delegator = std::make_shared<DB::tests::MockDiskDelegatorSingle>(path);
+        PageDirectoryFactory factory;
+        auto d = factory.createFromEdit(getCurrentTestName(), provider, delegator, deseri_edit);
         return d;
     };
 

--- a/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
@@ -132,33 +132,59 @@ TEST(WALSeriTest, Upserts)
     EXPECT_SAME_ENTRY(iter->entry, entry_p5_2);
 }
 
-TEST(WALSeriTest, RefExternal)
+TEST(WALSeriTest, RefExternalAndEntry)
 {
     PageVersionType ver1_0(/*seq=*/1, /*epoch*/ 0);
     PageVersionType ver2_0(/*seq=*/2, /*epoch*/ 0);
     PageVersionType ver3_0(/*seq=*/3, /*epoch*/ 0);
+    {
+        PageEntriesEdit edit;
+        edit.varExternal(1, ver1_0, 2);
+        edit.varDel(1, ver2_0);
+        edit.varRef(2, ver3_0, 1);
 
-    PageEntriesEdit edit;
-    edit.varExternal(1, ver1_0, 2);
-    edit.varDel(1, ver2_0);
-    edit.varRef(2, ver3_0, 1);
+        auto deseri_edit = DB::PS::V3::ser::deserializeFrom(DB::PS::V3::ser::serializeTo(edit));
+        ASSERT_EQ(deseri_edit.size(), 3);
+        auto iter = deseri_edit.getRecords().begin();
+        EXPECT_EQ(iter->type, EditRecordType::VAR_EXTERNAL);
+        EXPECT_EQ(iter->page_id.low, 1);
+        EXPECT_EQ(iter->version, ver1_0);
+        EXPECT_EQ(iter->being_ref_count, 2);
+        iter++;
+        EXPECT_EQ(iter->type, EditRecordType::VAR_DELETE);
+        EXPECT_EQ(iter->page_id.low, 1);
+        EXPECT_EQ(iter->version, ver2_0);
+        EXPECT_EQ(iter->being_ref_count, 1);
+        iter++;
+        EXPECT_EQ(iter->type, EditRecordType::VAR_REF);
+        EXPECT_EQ(iter->page_id.low, 2);
+        EXPECT_EQ(iter->version, ver3_0);
+    }
 
-    auto deseri_edit = DB::PS::V3::ser::deserializeFrom(DB::PS::V3::ser::serializeTo(edit));
-    ASSERT_EQ(deseri_edit.size(), 3);
-    auto iter = deseri_edit.getRecords().begin();
-    EXPECT_EQ(iter->type, EditRecordType::VAR_EXTERNAL);
-    EXPECT_EQ(iter->page_id.low, 1);
-    EXPECT_EQ(iter->version, ver1_0);
-    EXPECT_EQ(iter->being_ref_count, 2);
-    iter++;
-    EXPECT_EQ(iter->type, EditRecordType::VAR_DELETE);
-    EXPECT_EQ(iter->page_id.low, 1);
-    EXPECT_EQ(iter->version, ver2_0);
-    EXPECT_EQ(iter->being_ref_count, 1);
-    iter++;
-    EXPECT_EQ(iter->type, EditRecordType::VAR_REF);
-    EXPECT_EQ(iter->page_id.low, 2);
-    EXPECT_EQ(iter->version, ver3_0);
+    {
+        PageEntriesEdit edit;
+        PageEntryV3 entry_p1_2{.file_id = 2, .size = 1, .tag = 0, .offset = 0x123, .checksum = 0x4567};
+        edit.varEntry(1, ver1_0, entry_p1_2, 2);
+        edit.varDel(1, ver2_0);
+        edit.varRef(2, ver3_0, 1);
+
+        auto deseri_edit = DB::PS::V3::ser::deserializeFrom(DB::PS::V3::ser::serializeTo(edit));
+        ASSERT_EQ(deseri_edit.size(), 3);
+        auto iter = deseri_edit.getRecords().begin();
+        EXPECT_EQ(iter->type, EditRecordType::VAR_ENTRY);
+        EXPECT_EQ(iter->page_id.low, 1);
+        EXPECT_EQ(iter->version, ver1_0);
+        EXPECT_EQ(iter->being_ref_count, 2);
+        iter++;
+        EXPECT_EQ(iter->type, EditRecordType::VAR_DELETE);
+        EXPECT_EQ(iter->page_id.low, 1);
+        EXPECT_EQ(iter->version, ver2_0);
+        EXPECT_EQ(iter->being_ref_count, 1);
+        iter++;
+        EXPECT_EQ(iter->type, EditRecordType::VAR_REF);
+        EXPECT_EQ(iter->page_id.low, 2);
+        EXPECT_EQ(iter->version, ver3_0);
+    }
 }
 
 TEST(WALLognameTest, parsing)

--- a/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
@@ -132,6 +132,33 @@ TEST(WALSeriTest, Upserts)
     EXPECT_SAME_ENTRY(iter->entry, entry_p5_2);
 }
 
+TEST(WALSeriTest, RefExternal)
+{
+    PageVersionType ver1_0(/*seq=*/1, /*epoch*/ 0);
+    PageVersionType ver2_0(/*seq=*/2, /*epoch*/ 0);
+    PageVersionType ver3_0(/*seq=*/3, /*epoch*/ 0);
+
+    PageEntriesEdit edit;
+    edit.varExternal(1, ver1_0, 2);
+    edit.varDel(1, ver2_0);
+    edit.varRef(2, ver3_0, 1);
+
+    auto deseri_edit = DB::PS::V3::ser::deserializeFrom(DB::PS::V3::ser::serializeTo(edit));
+    ASSERT_EQ(deseri_edit.size(), 3);
+    auto iter = deseri_edit.getRecords().begin();
+    EXPECT_EQ(iter->type, EditRecordType::VAR_EXTERNAL);
+    EXPECT_EQ(iter->page_id.low, 1);
+    EXPECT_EQ(iter->version, ver1_0);
+    iter++;
+    EXPECT_EQ(iter->type, EditRecordType::VAR_DELETE);
+    EXPECT_EQ(iter->page_id.low, 1);
+    EXPECT_EQ(iter->version, ver2_0);
+    iter++;
+    EXPECT_EQ(iter->type, EditRecordType::VAR_REF);
+    EXPECT_EQ(iter->page_id.low, 2);
+    EXPECT_EQ(iter->version, ver3_0);
+}
+
 TEST(WALLognameTest, parsing)
 {
     LoggerPtr log = Logger::get("WALLognameTest");

--- a/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
@@ -149,10 +149,12 @@ TEST(WALSeriTest, RefExternal)
     EXPECT_EQ(iter->type, EditRecordType::VAR_EXTERNAL);
     EXPECT_EQ(iter->page_id.low, 1);
     EXPECT_EQ(iter->version, ver1_0);
+    EXPECT_EQ(iter->being_ref_count, 2);
     iter++;
     EXPECT_EQ(iter->type, EditRecordType::VAR_DELETE);
     EXPECT_EQ(iter->page_id.low, 1);
     EXPECT_EQ(iter->version, ver2_0);
+    EXPECT_EQ(iter->being_ref_count, 1);
     iter++;
     EXPECT_EQ(iter->type, EditRecordType::VAR_REF);
     EXPECT_EQ(iter->page_id.low, 2);

